esp/esdllib/esdl_script.cpp

```diff
@@ -1615,6 +1671,78 @@ class CEsdlTransformOperationTrace : public CEsdlTransformOperationWithoutChildr
         }
         return false;
     }
+
+    virtual bool getText(StringBuffer& text, IXpathContext* targetContext, IXpathContext* sourceContext, IDataMaskingProfileContext* masker) = 0;
+};
+
+class CEsdlTransformOperationTraceContent : public CEsdlTransformOperationTrace
+{
+protected:
+    StringAttr m_contentType;
+    Owned<ICompiledXpath> m_skipMask;
+
+public:
+    CEsdlTransformOperationTraceContent(IXmlPullParser &xpp, StartTag &stag, const StringBuffer &prefix) : CEsdlTransformOperationTrace(xpp, stag, prefix)
+    {
+        m_contentType.set(stag.getValue("content_type"));
+        const char* skipMask = stag.getValue("skip_mask");
+        if (!isEmptyString(skipMask))
+            m_skipMask.setown(compileXpath(skipMask));
+    }
+
+    virtual bool getText(StringBuffer& text, IXpathContext* targetContext, IXpathContext* sourceContext, IDataMaskingProfileContext* masker) override
+    {
+        bool isValue = false;
+        if (!targetContext->selectText(m_select, text, isValue) || !masker)
+            return true;
+        if (!m_skipMask || !sourceContext->evaluateAsBoolean(m_skipMask))
+        {
+            char* buffer = const_cast<char*>(text.str());
```

Comment:
I know that this is really the only realistic way to manage this but could we be opening ourselves up to problems by casting away the const? Are we guaranteed that masking will only ever change the contents and not the length?

Reply:
A cast seems to be necessary to avoid making copies of the data. We could define an encapsulation that prevents a plugin from directly manipulating the underlying character buffer, while also avoiding copies. Such an encapsulation would need to reinvent most (all) of the standard library string functionality in ways that don't expose the underlying buffer.

If it's important enough to do, it should probably be done as part of HPCC-27471 since it impacts the plugin interface. @afishbeck, @asselitx, thoughts?

Reply:
HPCC-27471 has now been merged...

Reply:
That is a big enough of a task I don't think it would make sense to do on this PR as you noted. I think Tony has a better sense of of how strongly this is discouraged. I do see a few other instances in the codebase.

Reply:
@ghalliday any thoughts on casting away const and passing the string to a mask routine that never changes the length, just masks characters?  TraceValue may be dealing with small strings, but TraceContent could potentially be large blobs.  For example some xml content.

Reply:
The concern, such as it is, is not introduced by this change. As mentioned in prior comments, it was introduced by HPCC-27471. There are three additional PRs building on this one that are unrelated to this concern.

If this warrants a change, I suggest it be done in a new JIRA focused on the change.

Reply:
Probably the cleanest way would be to add a member function to StringBuffer to return a char * e.g. bufferBase() used in MemoryBuffer.